firebase-firestore/src/androidTest/java/com/google/firebase/firestore/FirestoreTest.java

```diff
@@ -1493,4 +1495,157 @@ public void testCanGetSameOrDifferentPersistentCacheIndexManager() {
+            "__id12__", map("a", 1),
+            "__id-2__", map("a", 1),
+            "__id1_", map("a", 1),
+            "_id1__", map("a", 1),
+            "__id9223372036854775807__", map("a", 1),
+            "__id-9223372036854775808__", map("a", 1));
+
+    CollectionReference colRef = testCollectionWithDocs(testDocs);
+
+    // Run get query
+    Query filteredQuery =
+        colRef
+            .whereGreaterThan(FieldPath.documentId(), "__id7__")
+            .whereLessThanOrEqualTo(FieldPath.documentId(), "A")
+            .orderBy(FieldPath.documentId());
+    List<String> expectedDocIds =
+        Arrays.asList("__id12__", "__id9223372036854775807__", "12", "7", "A");
+
+    QuerySnapshot getSnapshot = waitFor(filteredQuery.get());
+    List<String> getSnapshotDocIds =
+        getSnapshot.getDocuments().stream().map(ds -> ds.getId()).collect(Collectors.toList());
+
+    // Run query with snapshot listener
+    EventAccumulator<QuerySnapshot> eventAccumulator = new EventAccumulator<QuerySnapshot>();
+    ListenerRegistration registration =
+        filteredQuery.addSnapshotListener(eventAccumulator.listener());
+
+    List<String> watchSnapshotDocIds = new ArrayList<>();
+    try {
+      QuerySnapshot watchSnapshot = eventAccumulator.await();
+      watchSnapshotDocIds =
+          watchSnapshot.getDocuments().stream()
+              .map(documentSnapshot -> documentSnapshot.getId())
+              .collect(Collectors.toList());
+    } finally {
+      registration.remove();
+    }
+
+    // Assert that get and snapshot listener requests sort docs in the same, expected order
+    assertTrue(getSnapshotDocIds.equals(expectedDocIds));
+    assertTrue(watchSnapshotDocIds.equals(expectedDocIds));
+  }
+
+  @Test
+  public void sdkOrdersQueryByDocumentIdTheSameWayOnlineAndOffline() {
+    Map<String, Map<String, Object>> testDocs =
+        map(
+            "A", map("a", 1),
+            "a", map("a", 1),
+            "Aa", map("a", 1),
+            "7", map("a", 1),
+            "12", map("a", 1),
+            "__id7__", map("a", 1),
+            "__id12__", map("a", 1),
+            "__id-2__", map("a", 1),
+            "__id1_", map("a", 1),
+            "_id1__", map("a", 1),
+            "__id9223372036854775807__", map("a", 1),
+            "__id-9223372036854775808__", map("a", 1));
+
+    CollectionReference colRef = testCollectionWithDocs(testDocs);
+    // Test query
+    Query orderedQuery = colRef.orderBy(FieldPath.documentId());
+    List<String> expectedDocIds =
+        Arrays.asList(
+            "__id-9223372036854775808__",
+            "__id-2__",
+            "__id7__",
+            "__id12__",
+            "__id9223372036854775807__",
+            "12",
+            "7",
+            "A",
+            "Aa",
+            "__id1_",
+            "_id1__",
+            "a");
+
+    // Run query with snapshot listener
+    checkOnlineAndOfflineResultsMatch(orderedQuery, expectedDocIds.toArray(new String[0]));
```

Comment:
I just realized we need to improve the the `checkOnlineAndOfflineResultsMatch` function (which I wrote a while back). That function uses `get(Source.Server)` and that does force the SDK to bypass the cache. However, the SDK sorts the results after it gets it from the server before presenting it to the user. So I think that function should be updated to use a transaction:

```java
  public static void checkOnlineAndOfflineResultsMatch(Query query, String... expectedDocs) {
    Task<QuerySnapshot> serverTransaction = query.getFirestore().runTransaction(transaction -> waitFor(query.get()));
    QuerySnapshot docsFromServer = waitFor(serverTransaction);
    QuerySnapshot docsFromCache = waitFor(query.get(Source.CACHE));

    assertEquals(querySnapshotToIds(docsFromServer), querySnapshotToIds(docsFromCache));
    List<String> expected = asList(expectedDocs);
    if (!expected.isEmpty()) {
      assertEquals(expected, querySnapshotToIds(docsFromCache));
    }
  }
}
```

Reply:
hmm... scratch that. It's actually not possible to run a query inside a transaction in mobile SDKs.